diff -rup a/localedata/SUPPORTED b/localedata/SUPPORTED
--- a/localedata/SUPPORTED	2012-03-12 14:11:04.134056609 -0600
+++ b/localedata/SUPPORTED	2012-03-14 20:59:39.993269235 -0600
@@ -103,6 +103,7 @@ de_DE@euro/ISO-8859-15 \
 de_LU.UTF-8/UTF-8 \
 de_LU/ISO-8859-1 \
 de_LU@euro/ISO-8859-15 \
+doi_IN/UTF-8 \
 dv_MV/UTF-8 \
 dz_BT/UTF-8 \
 el_GR.UTF-8/UTF-8 \
@@ -298,6 +299,7 @@ mk_MK.UTF-8/UTF-8 \
 mk_MK/ISO-8859-5 \
 ml_IN/UTF-8 \
 mn_MN/UTF-8 \
+mni_IN/UTF-8 \
 mr_IN/UTF-8 \
 ms_MY.UTF-8/UTF-8 \
 ms_MY/ISO-8859-1 \
@@ -350,6 +352,7 @@ ru_UA.UTF-8/UTF-8 \
 ru_UA/KOI8-U \
 rw_RW/UTF-8 \
 sa_IN/UTF-8 \
+sat_IN/UTF-8 \
 sc_IT/UTF-8 \
 sd_IN/UTF-8 \
 sd_IN@devanagari/UTF-8 \
