diff -rup c/time/sys/time.h d/time/sys/time.h
--- c/time/sys/time.h	2012-05-20 19:47:38.000000000 -0600
+++ d/time/sys/time.h	2012-05-21 00:14:17.033127079 -0600
@@ -77,7 +77,7 @@ extern int gettimeofday (struct timeval
    This call is restricted to the super-user.  */
 extern int settimeofday (const struct timeval *__tv,
 			 const struct timezone *__tz)
-     __THROW __nonnull ((1));
+     __THROW;
 
 /* Adjust the current time of day by the amount in DELTA.
    If OLDDELTA is not NULL, it is filled in with the amount
